File: Ch32/Rabin-Karp.lua
--Rabin-Karp-Matcher
-- based on number-theoretic principles

function Rabin_Karp_Matcher(Target,Pattern,prime)
  local Tlen=#Target
  local Plen=#Pattern
  local highbitpos=10^(Plen-1)%prime
  local Pmod=0
  local Tmod=0

  for i=1,Plen do
    Pmod = (10*Pmod+Pattern:sub(i,i))%prime
    Tmod =(10*Tmod +Target:sub(i,i))%prime
  end

  for shift=0,Tlen-Plen do
    if Pmod == Tmod then
      for i=0,Plen do
        if Pattern:sub(i,i)==Target:sub(i+shift,i+shift) then
	  return shift
	else
	  break
	end
      end

    if shift < Tlen-Plen then
      Tmod = (10*(Tmod-Target:sub(shift+1,shift+1)*highbitpos) +
             Target:sub(shift+i+Plen+1,shift+i+Plen+1))%prime

File: Ch32/rabinKarp.cpp
#include <iostream>
#include <string>

using namespace std;

int exp(int base,int e){
  int total=1;
  while (e>0){
    e -= 1;
    total *= base;
  }
  return total;
}
int rabinKarpMatcher(string T,string P,int q);

int main(){
  string T="2359023141526739921";
  string P="31415";
  int prime=13;
  cout<<"2^5 is "<<exp(2,5)<<endl;
  int result= rabinKarpMatcher(T,P,prime);
  cout<<"the result is "<<result<<endl;
  return 0;
}

int rabinKarpMatcher(string T,string P,int q){
  int n=T.length();
  int m=P.length();
  int h=exp(10,m-1) % q;
  int tmod=0;
  int pmod=0;
  //compute hash
  for(int i=0;i<m;i++){
    pmod=(10*pmod + P[i]) % q;
    tmod=(10*tmod + T[i]) % q;
  }
  for(int index=0;index < n-m; index++){
    if (pmod == tmod) {
      //this loop checks that the pattern exists char by char
      for(int shift=0;shift<m;shift++){
	if(P[index+shift] == T[index+shift]){
	  return index;
	}//innerif
      }//innerfor
    }//outerif
    if (index < n - m){
      tmod=(10*(tmod - T[index+1]*h) + T[index+m+1])%q;
    }
  }//outerfor
  return -1;
}
